hazelcast/include/hazelcast/client/sql/impl/read_optimized_lru_cache.h

```diff
@@ -0,0 +1,203 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <queue>
+#include "hazelcast/util/SynchronizedMap.h"
+#include "hazelcast/util/export.h"
+
+namespace hazelcast {
+namespace client {
+namespace sql {
+namespace impl {
+
+/**
+ * Implementation of an LRU cache optimized for read-heavy use cases.
+ * <p>
+ * It stores the entries in a {SynchronizedMap}, along with the last
+ * access time. It allows the size to grow beyond the capacity, up to
+ * `cleanup_threshold_`, at which point the inserting thread will remove a batch
+ * of the eldest items in two passes.
+ * <p>
+ * The cleanup process isn't synchronized to guarantee that the capacity is not
+ * exceeded. The cache is available during the cleanup for reads and writes. If
+ * there's a large number of writes by many threads, the one thread doing the
+ * cleanup might not be quick enough and there's no upper bound on the actual
+ * size of the cache. This is done to optimize the happy path when the keys fit
+ * into the cache.
+ */
+template<typename K, typename V>
+class read_optimized_lru_cache
+{
+public:
+    /**
+     * @param capacity Capacity of the cache
+     * @param cleanup_threshold The size at which the cache will clean up oldest
+     *    entries in batch. `cleanup_threshold - capacity` entries will be
+     * removed
+     * @throws exception::illegal_argument if capacity is smaller or equal to 0,
+     * or if the cleanup_threshold is smaller than capacity
+     */
+    explicit read_optimized_lru_cache(const uint32_t capacity,
+                                      const uint32_t cleanup_threshold)
+    {
+        if (capacity == 0) {
+            BOOST_THROW_EXCEPTION(
+              client::exception::illegal_argument("capacity == 0"));
+        }
+        if (cleanup_threshold <= capacity) {
+            BOOST_THROW_EXCEPTION(client::exception::illegal_argument(
+              "cleanupThreshold <= capacity"));
+        }
+
+        capacity_ = capacity;
+        cleanup_threshold_ = cleanup_threshold;
+    }
+
+    /**
+     * @param key the key of the cache entry
+     * @param default_value the default value if the key is not cached.
+     * @returns Returns the value to which the specified key is cached,
+     * or default value if this cache contains no mapping for the key.
+     */
+    std::shared_ptr<V> get_or_default(const K& key,
+                                      const std::shared_ptr<V>& default_value)
+    {
+        const auto existing_value = get(key);
+        return (existing_value != nullptr) ? existing_value : default_value;
+    }
+
+    /**
+     * @param key the key of the cache entry
+     * Returns the value to which the specified key is cached,
+     * or {@code null} if this cache contains no mapping for the key.
+     * @returns Returns the value to which the specified key is cached
+     */
+    std::shared_ptr<V> get(const K& key)
```

Comment:
why not return `optional` which makes small value optimizations for you?

Reply:
I made it to be consistent with SynchronizedMap, so I prefer to leave it as it is.

Reply:
Data is stored in synchronized map, and we just increment the usage count of the shared_ptr with get (which is done via atomic instruction). If we use optional, wouldn't it be expensive, because this time we will copy the data for storing it in optional?

Reply:
For this specific case, the value is an (int32 + int64_t) struct (`std::shared_ptr<impl::read_optimized_lru_cache<std::string, int32_t>>
      partition_argument_index_cache_;`) hence, copy would not be a big problem for this specific case. We have no other use of read_optimized_lru_cache structure any other places and have no such concern at the moment. Hence, I thought that it would have been better than using shared_ptr.